src/protagonist/Engine/Data/EngineAssetRepository.cs

```diff
@@ -163,10 +172,17 @@ private void UpdateAsset(Asset asset, bool ingestFinished)
         }
     }
 
-    private Task<int> TryFinishBatch(int batchId, CancellationToken cancellationToken)
-        => dlcsContext.Batches
-            .Where(b => b.Id == batchId && b.Count == b.Completed + b.Errors)
-            .UpdateFromQueryAsync(b => new Batch { Finished = DateTime.UtcNow }, cancellationToken);
+    private async Task<(int rowsAffected, IQueryable<Batch> batches)> TryFinishBatch(int batchId,
+        CancellationToken cancellationToken)
+    {
+        var batches = dlcsContext.Batches
+            .Where(b => b.Id == batchId && b.Count == b.Completed + b.Errors);
+        
+        var rowsAffected =
+            await batches.UpdateFromQueryAsync(b => new Batch { Finished = DateTime.UtcNow }, cancellationToken);

```

Comment:
The value in using `UpdateFromQueryAsync` is that you don't need to load any entities into context first, it just issues an update statement to DB directly. ie in 'before' diff it would be `update batches set finished = now() where id = @id and count = completed + errors`

However, this is now _maybe_ loading the batch but _always_ issuing the `update`. Might be cleaner written to always load the Batch, and let EF take care of working out whether to issue the update?

```cs
var batch = dlcsContext.Batches.Find(batchId);
// if batch is complete, set Finished
```

Reply:
Just thinking about this - the reason for issueing the UPDATE direct to the DB was to avoid multiple engine instances attempting to update the Batch at the same time. By the time instance 1 loads it instance 2 may have completed it. Is it better to issue update and only load the batch if rowsAffected as a value?

Reply:
That potentially feels safer, so I'll change it to only load if `rowsAffected > 0`